src/decisionengine/framework/engine/SourceWorkers.py

```diff
@@ -0,0 +1,186 @@
+                    self.logger.info(f"Source {self.name} acquire returned")
+                    SOURCE_ACQUIRE_GAUGE.labels(self.name).set_to_current_time()
+                    self.logger.debug(f"Publishing data to queue {self.key} with routing key {self.key}")
+                    producer.publish(
+                        dict(source_module=self.module, class_name=self.name, data=data),
+                        routing_key=self.key,
+                        exchange=self.exchange,
+                        serializer="pickle",
+                        declare=[
+                            self.exchange,
+                            self.queue,
+                        ],
+                    )
+                    self.logger.info(f"Source {self.name} {self.module} finished cycle")
+                except Exception:
+                    self.logger.exception(f"Exception running source {self.name} ")
+                    producer.publish(
+                        dict(source_module=self.module, class_name=self.name, data=State.SHUTDOWN),
+                        routing_key=self.key,
+                        exchange=self.exchange,
+                        serializer="pickle",
+                        declare=[
+                            self.exchange,
+                            self.queue,
+                        ],
+                    )
+                    break
+                if self.schedule > 0:
+                    time.sleep(self.schedule)
+                else:
+                    self.logger.info(f"Source {self.name} runs only once")
+                    break
+        self.logger.info(f"Stopped {self.name}")
+
+
+class SourceWorkers:
+    def __init__(self, exchange, broker_url, logger=structlog.getLogger(LOGGERNAME)):
+        self._exchange = exchange
+        self._broker_url = broker_url
+        self._logger = logger
+        self._workers = {}
+        self._lock = threading.Lock()
+
+    def update(self, channel_name, source_configs):
+        workers = {}
+
+        # Reuse already existing sources
+        with self._lock:
+            existing_sources = set(self._workers.keys()).intersection(source_configs.keys())
+            for src_name in existing_sources:
+                new_src_config = source_configs.pop(src_name)
+                src_worker = self._workers[src_name]
+                if new_src_config != src_worker.config:
+                    err_msg = (
+                        f"Channel {channel_name} will not be loaded due to the following error:\n"
+                        f"Mismatched configurations for source with name {src_name}\n"
+                        f"New configuration\n -> {new_src_config}\n"
+                        f"Cached configuration\n -> {src_worker.config}"
+                    )
+                    raise RuntimeError(err_msg)
+                self._logger.info(f"Using existing source {src_name} for channel {channel_name}")
+                workers[src_name] = src_worker
```

Comment:
Are the unit tests giving coverage here?

Reply:
Actually they're not.  Looks like I need to expand the unit tests a little.  Will do